aws/terraform/README.md

```diff
@@ -0,0 +1,112 @@
+## Pre-requisites
+
+### Terraform
+
+[Download](https://www.terraform.io/downloads.html) and install terraform 0.14.x.
+
+### AWS
+
+AWS should be [configured](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html) with a profile. The `AWS_PROFILE` environment should be set to the profile that has been configured.
+
+The [awscli](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html) can be used as follows to 
+confirm that it has been configured properly by running:
+
+```
+aws sts get-caller-identity
+```
+
+which should output your account information.
+
+## Setup
+
+### Infrastructure stack
+
+The infra sub-project provides some pre-requisite infrastructure for the metaflow service. For more details see the [README](aws/terraform/infra/README.md)
+
+Copy `example.tfvars` to `prod.tfvars` (or whatever environment name you prefer) and update that `env` name and the `region` as needed.
+
+Initialize the terraform:
+
+`cd infra && terraform init`
+
+Apply it:
+
+```
+terraform apply --var-file prod.tfvars
+```
+
+### Metaflow stack
+
+The metaflow sub-project provisions the metadata API, step functions, and an AWS Batch queue. For more details see the 
+[README](aws/terraform/metaflow/README.md)
+
+Copy `example.tfvars` to `prod.tfvars` (or whatever environment name you prefer) and update that `env` name and the `region` as needed.
+Metadata API authentication: the endpoint is exposed to the public internet via API Gateway, but only accessible to the IPs that match `access_list_cidr_blocks` (default is none).
+Additionally:
+* There are variables which govern the three compute environments associated with the AWS Batch queue that can be adjusted based on needs. 
```

Comment:
Why are three batch compute environments needed?

Reply:
That's a great question! They aren't necessarily needed but we found ourselves, when testing our Metaflow, following into a pattern of separating our work into three compute environments that we thought would work with most people's workflows and end up saving them both headache and infrastructure costs. I'll document this "pattern" here and then update the documentation to reflect this at `aws/terraform/modules/metaflow/modules/computation/README/.md`.

At a minimum, all one really needs is a single AWS Batch Compute Environment to run Metaflow. We found this to be convenient for our Data Scientists, as we were able to add all the EC2 types we'd possibly need to this Compute Environment and the Data Scientists would not have to decide where to run their job, it would just run.

What we ended up discovering is that AWS Batch will automatically scale down smaller instances to make room for larger instances if a particularly large job was submitted, which is a good thing. However what could happen is these particularly large instances could hang around after the occasional large job was processed, as long as instance still respected the Compute Environment's min/max vCPU. So we ended up finding our Compute Environments, that often could run with say a `t2.medium` housing many GPU instances as a few occasional GPU jobs came through. 

To resolve this we ended up breaking our single AWS Batch Compute Environment up into three separate ones with very specific use cases:

1. `cpu`: Handles most small-medium jobs.
2. `large-cpu`: Handles large CPU only jobs.
3. `gpu`: Handles GPU jobs.

This allowed us to set our min/max vCPU on each Compute Environment separately which allowed us to force `large-cpu` and `gpu` to scale down automatically, while ensuring our `cpu` Compute Environment was both warm and not wasting idle computation costs.

Does that help explain the scenario? Any thoughts or questions?

@kldavis4 Anything you'd like to add?

Reply:
At this moment, I'm working on a change to our local metaflow terraform module that will allow specification via variables the job queues and compute environments to make this more flexible. If we were to do the same thing here, then as a default, we could just have a single queue with a single environment, but consumers could setup more sophisticated batch compute environments.

Reply:
I understand the desire to simplify to the bare minimum, but as a data-point on the other side -- we have had to implement this (or a fairly similar) pattern on other Metaflow deployments as well. 

I'll defer to others on the final call, but this was a hard-learned lesson re: how Metaflow interacts with AWS Batch and I hope we can streamline other folks adoption by simplifying the path to a minimal "useful" deployment.

Reply:
For an example of what we've done in our version of the metaflow module configuration, we've dropped the min/max/desired vcpu vars for the three environments and now specify a batch_job_queues configuration. To replicate the current infrastructure then becomes:

```
job_queues = {
  default = {
    compute_environments = {
      cpu = {
        instance_types = [
          "r5.large",
          "r5.xlarge",
          "r5.2xlarge",
          "c4.large",
          "c4.xlarge",
          "c4.2xlarge",
          "c4.4xlarge"
        ]

        max_vcpus     = 128
        min_vcpus     = 16
        desired_vcpus = 16
      }

      large-cpu = {
        instance_types = [
          "c4.8xlarge",
          "r5.4xlarge",
          "r5.8xlarge",
          "r5.12xlarge"
        ]

        max_vcpus     = 128
        min_vcpus     = 0
        desired_vcpus = 0
      }

      gpu = {
        instance_types = [
          "p3.2xlarge"
        ]

        max_vcpus     = 128
        min_vcpus     = 0
        desired_vcpus = 0

        template = {
          ami_type = "gpu"
        }
      }
    }
  }
```

We could then simplify the default somewhat (dropping large-cpu). Because we are using (as of 96b15df) the amazon linux 2 ecs optimized images, though, I think we will need at least 2 compute environments in order to support both cpu and gpu. In my tests, trying to use gpu instances with the cpu version of AL2 caused issues.